Verify revision identity before scale-up E2E test

<!--
/area autoscale
/area test-and-release
/kind bug
-->

## Expected Behavior

The `AutoscaleUpDownUp` test should pass reliably.  In particular the first scale up should always work.

## Actual Behavior

There are periodic flakes in which scale-up just doesn't happen.

Sometimes there are multiple services deployed and some have multiple revisions.  It is not clear which revision is under test.  And we do not verify that we are hitting the exact revision we are looking for.  E.g. traffic might be going to one revision while we wait for another to scale up !?

We should:
1. Log the revision name under test (just for clarity).
2. Inject a new, random id into the revision under test and verify that unique id is consistently being returned as a precondition for starting the first scale-up.

https://github.com/knative/serving/blob/ad10693a0a766ab8396048c970cefb1e96bf709e/test/e2e/autoscale_test.go#L184

## Additional Info

Here is an example of a flake (note the `diagnose.go` output which shows multiple revisions):
```
I1024 07:16:50.994] 2018-10-24T07:13:36.723Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:209	The autoscaler spins up additional replicas when traffic increases.
I1024 07:16:50.994] 2018-10-24T07:13:36.723Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:70	Maintaining 20 concurrent requests for 20s.
I1024 07:16:50.994] 2018-10-24T07:13:36.725Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:115	Waiting for all requests to complete.
I1024 07:16:50.994] 2018-10-24T07:13:36.793Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodgnfuaoex-00001-deployment has 1 pods.
I1024 07:16:50.995] 2018-10-24T07:13:36.800Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodgnfuaoex-00002-deployment has 1 pods.
I1024 07:16:50.995] 2018-10-24T07:13:36.806Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:50.995] 2018-10-24T07:13:37.738Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:98	Request took: 1.003419419s
I1024 07:16:50.995] 2018-10-24T07:13:37.749Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:98	Request took: 1.008465796s
I1024 07:16:50.995] 2018-10-24T07:13:37.749Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:98	Request took: 1.013046267s
...
I1024 07:16:51.020] 2018-10-24T07:13:54.157Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodbbkvveyb-00001-deployment has 1 pods.
I1024 07:16:51.020] 2018-10-24T07:13:54.238Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:59	deployment prodbbkvveyb-00002-deployment has 0 pods. wants 1.
I1024 07:16:51.021] 2018-10-24T07:13:54.321Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.021] 2018-10-24T07:13:54.754Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:98	Request took: 2.001382224s
I1024 07:16:51.021] 2018-10-24T07:13:54.764Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:98	Request took: 2.00156503s
I1024 07:16:51.021] 2018-10-24T07:13:54.771Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:98	Request took: 2.001807707s
...
I1024 07:16:51.027] 2018-10-24T07:13:57.828Z	info	TestAutoscaleUpDownUp	e2e/autoscale_test.go:214	Waiting for scale up
I1024 07:16:51.027] 2018-10-24T07:14:09.383Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment pizzaplanet-servicesajjuliw-00001-deployment has 1 pods.
I1024 07:16:51.027] 2018-10-24T07:14:09.406Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment pizzaplanet-servicesajjuliw-00002-deployment has 1 pods.
I1024 07:16:51.027] 2018-10-24T07:14:09.431Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.028] 2018-10-24T07:14:24.479Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.028] 2018-10-24T07:14:39.523Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.028] 2018-10-24T07:14:54.568Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.028] 2018-10-24T07:15:09.589Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.028] 2018-10-24T07:15:24.611Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.028] 2018-10-24T07:15:39.630Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.029] 2018-10-24T07:15:54.653Z	info	TestAutoscaleUpDownUp	e2e/diagnose.go:61	deployment prodnrzlndni-00001-deployment has 1 pods.
I1024 07:16:51.029] 2018-10-24T07:15:57.874Z	info	TestAutoscaleUpDownUp	trace/trace.go:229	metric WaitForDeploymentState/prodnrzlndni-00001-deployment/DeploymentIsScaledUp 1540365237828534902 1540365357874586968 2m0.046052066s
I1024 07:16:51.029] --- FAIL: TestAutoscaleUpDownUp (180.56s)
```
([source](https://gubernator.knative.dev/build/knative-prow/logs/ci-knative-serving-continuous/1054991360009965568?log#log))

--- a/test/e2e/autoscale_test.go
+++ b/test/e2e/autoscale_test.go
@@ -209,7 +209,6 @@ func assertScaleUp(ctx *testContext) {
 	if err != nil {
 		ctx.t.Fatalf("Unable to observe the Deployment named %s scaling up. %s", ctx.deploymentName, err)
 	}
-
 }
 
 func assertScaleDown(ctx *testContext) {

--- a/test/e2e/autoscale_test.go
+++ b/test/e2e/autoscale_test.go
@@ -210,7 +210,7 @@ func assertScaleUp(ctx *testContext) {
 	}
 
 	// verify scale-up has happened
-	if dep.Status.ReadyReplicas <= 0 {
+	if dep.Status.ReadyReplicas == 0 {
 		ctx.t.Fatalf("Deployment %s failed to scale up", ctx.deploymentName)
 	}
 

--- a/test/e2e/autoscale_test.go
+++ b/test/e2e/autoscale_test.go
@@ -177,6 +177,12 @@ func setup(t *testing.T) *testContext {
 	}
 
 	t.Logf("Revision under test: %s", names.Revision)
+
+	// verify that revision is consistent and is ready
+	if err := test.CheckRevisionState(clients.ServingClient, names.Revision, test.IsRevisionReady); err != nil {
+		t.Fatalf("The Revision %s is not in ready state: %v", names.Revision, err)
+	}
+
 	return &testContext{
 		t:              t,
 		clients:        clients,
@@ -202,16 +208,6 @@ func assertScaleUp(ctx *testContext) {
 		2*time.Minute)
 	if err != nil {
 		ctx.t.Fatalf("Unable to observe the Deployment named %s scaling up. %s", ctx.deploymentName, err)
-	}
-
-	dep, err := ctx.clients.KubeClient.Kube.AppsV1().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
-	if err != nil {
-		ctx.t.Fatalf("Could not get deployment %v", ctx.deploymentName)
-	}
-
-	// verify scale-up has happened
-	if dep.Status.ReadyReplicas == 0 {
-		ctx.t.Fatalf("Deployment %s failed to scale up", ctx.deploymentName)
 	}
 
 }
@@ -250,7 +246,7 @@ func assertScaleDown(ctx *testContext) {
 
 	ctx.t.Log("The Revision should remain ready after scaling to zero.")
 	if err := test.CheckRevisionState(ctx.clients.ServingClient, ctx.names.Revision, test.IsRevisionReady); err != nil {
-		ctx.t.Fatalf("The Revision %s did not stay Ready after scaling down to zero: %v", ctx.names.Revision, err)
+		ctx.t.Fatalf("The Revision %s did not stay Ready after scaling down to zerozero: %v", ctx.names.Revision, err)
 	}
 
 	ctx.t.Log("Scaled down.")

--- a/test/e2e/autoscale_test.go
+++ b/test/e2e/autoscale_test.go
@@ -246,7 +246,7 @@ func assertScaleDown(ctx *testContext) {
 
 	ctx.t.Log("The Revision should remain ready after scaling to zero.")
 	if err := test.CheckRevisionState(ctx.clients.ServingClient, ctx.names.Revision, test.IsRevisionReady); err != nil {
-		ctx.t.Fatalf("The Revision %s did not stay Ready after scaling down to zerozero: %v", ctx.names.Revision, err)
+		ctx.t.Fatalf("The Revision %s did not stay Ready after scaling down to zero: %v", ctx.names.Revision, err)
 	}
 
 	ctx.t.Log("Scaled down.")

--- a/test/e2e/autoscale_test.go
+++ b/test/e2e/autoscale_test.go
@@ -198,7 +198,7 @@ func assertScaleUp(ctx *testContext) {
 	if err != nil {
 		ctx.t.Fatalf("Error during initial scale up: %v", err)
 	}
-	ctx.t.Logf("Waiting for scale up revsion %s", ctx.names.Revision)
+	ctx.t.Logf("Waiting for scale up revision %s", ctx.names.Revision)
 	err = pkgTest.WaitForDeploymentState(
 		ctx.clients.KubeClient,
 		ctx.deploymentName,

--- a/test/e2e/autoscale_test.go
+++ b/test/e2e/autoscale_test.go
@@ -176,6 +176,7 @@ func setup(t *testing.T) *testContext {
 			names.Route, domain, autoscaleExpectedOutput, err)
 	}
 
+	t.Logf("Revision under test: %s", names.Revision)
 	return &testContext{
 		t:              t,
 		clients:        clients,
@@ -191,7 +192,7 @@ func assertScaleUp(ctx *testContext) {
 	if err != nil {
 		ctx.t.Fatalf("Error during initial scale up: %v", err)
 	}
-	ctx.t.Log("Waiting for scale up")
+	ctx.t.Logf("Waiting for scale up revsion %s", ctx.names.Revision)
 	err = pkgTest.WaitForDeploymentState(
 		ctx.clients.KubeClient,
 		ctx.deploymentName,
@@ -202,6 +203,17 @@ func assertScaleUp(ctx *testContext) {
 	if err != nil {
 		ctx.t.Fatalf("Unable to observe the Deployment named %s scaling up. %s", ctx.deploymentName, err)
 	}
+
+	dep, err := ctx.clients.KubeClient.Kube.AppsV1().Deployments(test.ServingNamespace).Get(ctx.deploymentName, metav1.GetOptions{})
+	if err != nil {
+		ctx.t.Fatalf("Could not get deployment %v", ctx.deploymentName)
+	}
+
+	// verify scale-up has happened
+	if dep.Status.ReadyReplicas <= 0 {
+		ctx.t.Fatalf("Deployment %s failed to scale up", ctx.deploymentName)
+	}
+
 }
 
 func assertScaleDown(ctx *testContext) {

